lib/src/import_path_base.dart

```diff
@@ -0,0 +1,283 @@
+      }
+
+      if (findAll) {
+        printMessage(
+            '» Searching for all import paths for `$importToFind`...\n');
+      } else {
+        printMessage(
+            '» Searching for the shortest import path for `$importToFind`...\n');
+      }
+    }
+
+    var tree = await search(dots: dots);
+
+    if (tree != null) {
+      var treeText = tree.generate();
+      printMessage(treeText);
+    }
+
+    if (!quiet) {
+      if (tree == null) {
+        printMessage(
+            '» Unable to find an import path from $from to $importToFind');
+      } else {
+        var totalFoundPaths = tree.totalLeafs;
+        if (totalFoundPaths > 1) {
+          printMessage(
+              '» Found $totalFoundPaths import paths from $from to $importToFind\n');
+        }
+      }
+    }
+
+    return tree;
+  }
+
+  /// Performs the imports search and returns the tree.
+  /// - If [dots] is `true` uses the `dots` style for the tree.
+  /// - See [ASCIIArtTree].
+  Future<ASCIIArtTree?> search({bool dots = false}) async {
+    var foundPaths = await searchPaths();
+    if (foundPaths == null || foundPaths.isEmpty) return null;
+
+    var asciiArtTree = ASCIIArtTree.fromPaths(
+      foundPaths,
+      stripPrefix: stripSearchRoot,
+      style: dots ? ASCIIArtTreeStyle.dots : ASCIIArtTreeStyle.elegant,
+    );
+
+    return asciiArtTree;
+  }
+
+  late PackageConfig _packageConfig;
+  late String _generatedDir;
+
+  /// Performs the imports search and returns the found import paths.
+  Future<List<List<String>>?> searchPaths() async {
+    var currentDir = Directory.current;
+    _packageConfig = (await findPackageConfig(currentDir))!;
+    _generatedDir = p.join('.dart_tool/build/generated');
+
+    var root = from.scheme == 'package' ? _packageConfig.resolve(from)! : from;
+
+    var foundPaths = _searchImportPaths(root, stripSearchRoot: stripSearchRoot);
+    if (foundPaths.isEmpty) {
+      return null;
+    }
+
+    if (!findAll) {
+      foundPaths.sort((a, b) => a.length.compareTo(b.length));
+      var shortest = foundPaths.first;
+      return [shortest];
+    } else {
+      return foundPaths;
+    }
+  }
+
+  List<List<String>> _searchImportPaths(
+    Uri node, {
+    String? stripSearchRoot,
+    Set<Uri>? walked,
+    List<String>? parents,
+    List<List<String>>? found,
+  }) {
+    found ??= [];
+
+    if (walked == null) {
+      walked = {node};
+    } else if (!walked.add(node)) {
+      return found;
+    }
+
+    final nodePath = node.toString();
+    if (parents == null) {
+      parents = [nodePath];
+    } else {
+      parents.add(nodePath);
+    }
+
+    var newImports = _importsFor(node, quiet: quiet)
+        .where((uri) => !walked!.contains(uri))
+        .toList(growable: false);
+
+    for (var import in newImports) {
+      if (_matchesImport(importToFind, import)) {
+        var foundPath = [...parents, import.toString()];
+        found.add(foundPath);
```

Comment:
Consider creating a ChildNode type, which has a pointer to its parent node (but not the other way).  This should avoid the need for lists entirely for the paths, which will avoid extra copying around like this. You could instead return iterables, since it's easy to crawl from a child up to all its parents.

We could actually represent a more complex graph that way too, if we allowed multiple parents. And possibly have a mode where we would serve a visual graph or output a graph in dot format? (not for now, but could be a nice follow up).

A representation like that would also resolve the issue I described above, finding an already "walked" node would actually mean just adding the current node to its parents.

Reply:
Fixed with the use of package graph_explorer

Reply:
Fwiw, I tested this using my fake package generator tool, which basically just generates packages with really complex import graphs for worst case scenarios. It seems to perform just fine in the shortest path mode 👍.

When there are many paths the `--all` mode definitely struggles a lot (took several minutes in my case, I gave up after 5). I think there could definitely be a lot of optimization there - ultimately the paths are still being represented with separate lists from what I could tell, which involves a huge amount of duplicate work. They should be represented as Iterables instead, which can be lazily printed and constructed.

I think the model for output in that mode should also really be a graph format (something like the DOT language for graphviz), and we could even then do something similar to the [pubviz](https://pub.dev/packages/pubviz) package does, creating/serving an HTML page for viewing in a browser. The console is just really not a good mechanism for efficiently visualizing large graphs 🤣. In general this format would also be more useful for tools than JSON (for the same reasons, orders of magnitude smaller representation).

I could also envision possibly some sort of interactive console mode, where you can say "show me the next path". If we can lazily materialize the paths, then we can always only show the first X (maybe 10?) and then have a sort of paginated setup where you can ask for more. That way it won't just hang forever, or spit out an infinite stream of crud to the screen :).

This should all be a follow up if anything, for many cases the existing `--all` implementation will probably work well enough. But I do think it should possibly print a warning that on large programs it might not complete in reasonable time.

Reply:
Can I test your generator? So I can try to find the bottleneck in the code, beside that parents list?

Reply:
https://gist.github.com/jakemac53/890f7dac88419f176e964458f9a12f7e

But, this is a fundamental flaw that you won't be able to optimize yourself out of I am quite confident :).